include/rapidcheck/state/Commands.hpp

```diff
@@ -1,9 +1,155 @@
+#include <mutex>
+#include "rapidcheck/detail/Barrier.h"
 
 namespace rc {
 namespace state {
+namespace detail {
+
+template <typename Model, typename Cmd>
+void allInterleavingsAreValid(const Commands<Cmd> &left,
+                              const Commands<Cmd> &right,
+                              int leftIndex,
+                              int rightIndex,
+                              const Model &state) {
+  const auto hasLeft = leftIndex < left.size();
+  const auto hasRight = rightIndex < right.size();
+
+  if (hasLeft) {
+    auto preState = state;
+    auto currentState = state;
+    auto command = left[leftIndex];
+    command->apply(currentState);
+    allInterleavingsAreValid(
+        left, right, leftIndex + 1, rightIndex, currentState);
+  }
+
+  if (hasRight) {
+    auto preState = state;
+    auto currentState = state;
+    auto command = right[rightIndex];
+    command->apply(currentState);
+    allInterleavingsAreValid(
+        left, right, leftIndex, rightIndex + 1, currentState);
+  }
+}
+
+template <typename Model, typename Cmd>
+void isValidSequence(
+  const ParallelCommands<Cmd> &cmds,
+  const Model &state) {
+
+  auto currentState = state;
+
+  // Check prefix
+  for (const auto &command : cmds.prefix) {
+    command->apply(currentState);
+  }
+
+  // verify parallel sequences
+  allInterleavingsAreValid(
+    cmds.left, cmds.right, 0, 0, currentState);
+}
+
+
+template <typename Model, typename Cmd>
+struct CommandResult
+{
+  CommandResult(const std::shared_ptr<const Cmd>& command,
+                std::function<void(const Model&)>&& verifyFunc)
+    : command(command)
+    , verifyFunc(std::move(verifyFunc))
+  {
+  }
+
+  std::shared_ptr<const Cmd> command;
+  std::function<void(const Model&)> verifyFunc;
+};
+
+template <typename Model, typename Cmd>
+struct ParallelExecutionResult {
+  std::vector<CommandResult<Model, Cmd>> prefix;
+  std::vector<CommandResult<Model, Cmd>> left;
+  std::vector<CommandResult<Model, Cmd>> right;
+};
+
+
+template <typename Model, typename Cmd>
+bool hasValidInterleaving(const std::vector<CommandResult<Model, Cmd>> &left,
+                          const std::vector<CommandResult<Model, Cmd>> &right,
+                          int leftIndex,
+                          int rightIndex,
+                          const Model &state) {
+  const auto hasLeft = leftIndex < left.size();
+  const auto hasRight = rightIndex < right.size();
+
+  // Base case, no more commands to verify
+  if (!hasRight && !hasLeft) {
+    return true;
+  }
+
+  auto isValidLeft = false;
+  auto isValidRight = false;
+
+  if (hasLeft) {
+    try {
+      auto preState = state;
+      auto currentState = state;
+      auto commandResult = left[leftIndex];
+      commandResult.command->apply(currentState);
+      commandResult.verifyFunc(preState);
+      isValidLeft = hasValidInterleaving(
+          left, right, leftIndex + 1, rightIndex, currentState);
+    } catch (...) {
+      // Failed
+    }
+  }
```

Comment:
Return true at the end of each `has___` branch and return false at the bottom of the method if both fall through instead. That way if the first branch has a valid interleaving, we don't check the second one unnecessarily.